14章&15章解きました & レビューお願いします

sukechannnn opened the pull request:
- たのしいRubyの14章&15章の練習問題を解きました

sukechannnn committed:
14章&15章書いた & レビューお願いします

hanocha commented on exercises_14_15.rb:
```diff
@@ -0,0 +1,99 @@
+# Chapter 14
+
+module Chapter14
+  class Exercise1
```
split の引数に何も渡さないと `' '` で区切ってくれる、という豆知識。
あ、明示的に書くことは悪いことじゃないので、あくまで豆知識として覚えておけばいいと思います！

あと、今回の問題の趣旨とは違うかもしれないけど、半角スペース区切りの文字列を配列化するなら以下のような書き方をよく使うかな。(知ってたらごめんね)

```
%w(Ruby is an object oriented programming language)
```

hanocha commented on exercises_14_15.rb:
```diff
@@ -0,0 +1,99 @@
+  class Exercise2
+    def self.sort_by_alphabet
+      ary = Exercise1.string_to_ary
+      ary.sort_by { |a| a[0] }
```
単純な sort でもよさそうだけど、わざわざ sort_by にした理由はなんだろう？

sukechannnn replied:
その方が高速だからです！

sukechannnn replied:
と思ったのですが、`sort`のみでも良かったですね。

hanocha commented on exercises_14_15.rb:
```diff
@@ -0,0 +1,99 @@
+
+  class Exercise4
+    def self.count_char
+      char_ary = 'Ruby is an object oriented programming language'.split('')
```
`each_char` メソッドがあるから、わざわざ split('') で配列を作らなくてもいいかも？

sukechannnn replied:
なるほど

sukechannnn replied:
配列作らずにソートするの、どうしたら良いか思いつかないんですが、なにか良い書き方ありますか？

hanocha replied:
ぱっと今コード出てこないけど、多少行数多くなってもいい、という前提なら書けそう？

hanocha replied:
一回配列にしちゃったほうがソートは面倒くさくないか。今回はこのままでも良いかも。ｽﾏﾝ 🙇 

あとは好みの問題だけど、ソートは配列作成時にやっちゃったほうが読みやすいかもしれないとは思った。

```
char_ary = 'Ruby is an object oriented programming language'.split('').sort
char_hash = char_ary.each_with_object(Hash.new(0)) { |a, hash| hash[a] += 1; }
```

hanocha commented on exercises_14_15.rb:
```diff
@@ -0,0 +1,99 @@
+    def self.count_char
+      char_ary = 'Ruby is an object oriented programming language'.split('')
+      char_hash = char_ary.sort.each_with_object(Hash.new(0)) { |a, hash| hash[a] += 1; }
+      char_hash.each { |k, v| print "'#{k}': "; v.times { print '*' }; puts }
```
短く書こうとして色々工夫してる、という印象を受けたけど、そのせいで可読性が犠牲になってる感じがする。
28行目とか、以下のように書くと何をしているかが分かりやすくなると思うんだけど、どうだろう？

```
char_hash.each do |char, count|
  puts "'#{char}': #{'*' * count}"
end
```

char_hash を char_counts とかにしてもいいかも。各文字のカウントが入ってるよ、的な意味で。

sukechannnn replied:
> char_hash を char_counts とかにしてもいいかも。各文字のカウントが入ってるよ、的な意味で。

文字のカウントが入ってるっていうよりもただソートした文字が入っているだけなんですが、そういうコンテキストを変数名にしてしまって良いのでしょうか？

hanocha replied:
あれ、文字とその出現回数のハッシュになってたと思ったんだけど違ったっけ？

sukechannnn replied:
> 文字とその出現回数のハッシュになってた

あ、そうですね！
そうしたら確かにメソッド名変えた方が良さそうですね。

hanocha commented on exercises_14_15.rb:
```diff
@@ -0,0 +1,99 @@
+    KANNUM_10 = { '千' => 1000, '百' => 100, '十' => 10, '' => 1 }.freeze
+
+    def self.kan2num(kan)
+      kan.scan(/([^千百十]*)([千百十]?)/).inject(-1) do |ret, (first_digit, unit)|
```
`kan.scan(/([^千百十]*)([千百十]?)/)` で、数値を取り出してる部分がなんで `*` なのかが気になった。
`?` じゃダメなんだろうか？桁指定の方は `?` だけど、一緒にしてないのはなんでだろうと思ったので。
（`/([^千百十]?)([千百十]?)` にしても動くのは確認した。）

sukechannnn replied:
`?`で良いですね。
ありがとうございます！

hanocha commented on exercises_14_15.rb:
```diff
@@ -0,0 +1,99 @@
+
+  class Exercise3
+    def self.exercise3
+      Chapter15::Exercise1.wday.each { |k, v| puts "「#{k}」は#{v}のことです。" }
```
いっそ k, v を en, jp とかにしたら分かりやすいかも？とか思ったけど、
このコードでそこまですることはないか、という気持ち。

もっと複雑な（大規模な）コード中で同じようなことをするなら、上記みたいな工夫してもいいかも、とか思った。

sukechannnn replied:
> もっと複雑な（大規模な）コード中で同じようなことをするなら

そうですね、今後はそういう想定でコード書いてみます！

hanocha commented:
「たのしいRuby」をやる目的は基礎の確認、復習なんだけど、
知識としては結構知ってることも多そうだし、読みやすさ、レビューしやすさみたいなところを意識して解答を作ってもらうとより良いかも？と思いました！

sukechannnn committed:
renamed:    exercises_14_15.rb -> chapter_14_15.rb

sukechannnn committed:
コメント頂いた部分を修正&改善しました

- 14章の(4)をシンプルにして、ブロックをdo ~
endにすることで可読性を上げる
- *は過剰なので変更 * -> ?
- 可読性のためにリファクタリング

sukechannnn committed:
modified sort_by { ~ } -> sort

hanocha commented:
https://github.com/yosuke0315/tanoshii_ruby_practice/pull/1#discussion_r113090058

ここだけちょっと気になってるので、これだけ回答もらえればと思います！

sukechannnn committed:
変数名をより適切なものに変更

char_hash -> char_count

yosuke0315 commented:
ありがとうございます！
修正しました!

hanocha commented:
LGTM! マージします！

![LGTM](http://forgifs.com/gallery/d/225853-4/Parrot-running-explosions.gif?)

hanocha commented:
あ、マージできないじゃんこれ…あれか、コラボレーターにならないといけない系のか

yosuke0315 commented:
なるほど！
招待しますね！
